File: packages/dataframe/rust/stats/distributions/negative_binomial.test.R
#!/usr/bin/env Rscript

# Negative binomial distribution test script
# Usage: Rscript negative_binomial.test.R <function> <x_or_p> <r> <p> [lower_tail] [log_p]

source("test-helpers.R")

# Define negative binomial distribution functions
negative_binomial_density <- function(x, r, p, log = FALSE) {
  dnbinom(x, r, p, log = log)
}

negative_binomial_cumulative <- function(x, r, p, lower.tail = TRUE, log.p = FALSE) {
  pnbinom(x, r, p, lower.tail = lower.tail, log.p = log.p)
}

negative_binomial_quantile <- function(p, r, prob, lower.tail = TRUE, log.p = FALSE) {
  qnbinom(p, r, prob, lower.tail = lower.tail, log.p = log.p)
}

negative_binomial_random <- function(n, r, p) {
  rnbinom(n, r, p)
}

# Run the test
run_distribution_test(negative_binomial_density, negative_binomial_cumulative, negative_binomial_quantile, negative_binomial_random)

File: packages/dataframe/rust/stats/distributions/negative_binomial.test.ts
#!/usr/bin/env -S deno run --allow-read --allow-run

import { callR, printResults, runComparison } from "./test-helpers.ts";

async function main() {
  console.log("Testing R vs Rust negative binomial functions...\n");

  const testCases = [
    {
      func: "dnbinom",
      distribution: "negative_binomial",
      args: [2, 5, 0.3, 0],
    },
    {
      func: "pnbinom",
      distribution: "negative_binomial",
      args: [2, 5, 0.3, 1, 0],
    },
    {
      func: "qnbinom",
      distribution: "negative_binomial",
      args: [0.0287955, 5, 0.3, 1, 0],
    },
  ];

  const results = await runComparison(testCases);
  printResults(results);

  // Test rnbinom separately since it's not available in Rust interface
  console.log("\nR-only tests:");
  try {
    const rnbinomResult = await callR(
      "rnbinom",
      "negative_binomial",
      1,
      5,
      0.3,
    );
    console.log(`rnbinom(5, 0.3) = ${rnbinomResult}`);
  } catch (error) {
    console.log(`rnbinom: Error - ${(error as Error).message}`);
  }
}

if (import.meta.main) {
  main().catch(console.error);
}
